File: usertests/kvm/host/src/vcpu/pagetable.cpp
#include "vcpu.hpp"

#include "bus.hpp"

#include <cstdio>
#include <cstring>
#include <limits>

namespace kvm_host {
static constexpr std::uintptr_t PAGE_SHIFT = 12;
static constexpr std::uintptr_t PAGE_SIZE = 1 << PAGE_SHIFT;
static constexpr std::uintptr_t PTE_SIZE = 8;
static constexpr std::uintptr_t VPN_BITS = 9;
static constexpr std::uintptr_t VPN_MASK = (1 << VPN_BITS) - 1;
static constexpr std::uintptr_t SATP_MODE_BARE = 0;
static constexpr std::uintptr_t SATP_MODE_SV39 = 8;
static constexpr std::uintptr_t SATP_MODE_SHIFT = 60;
static constexpr std::uintptr_t SATP_PPN_MASK = (static_cast<std::uintptr_t>(1) << 44) - 1;

static constexpr std::uint64_t PTE_V = 1 << 0;
static constexpr std::uint64_t PTE_R = 1 << 1;
static constexpr std::uint64_t PTE_W = 1 << 2;
static constexpr std::uint64_t PTE_X = 1 << 3;
static constexpr std::uint64_t PTE_PPN_MASK = (static_cast<std::uint64_t>(1) << 44) - 1;

static bool checked_range_end(std::uintptr_t start, std::uintptr_t length, std::uintptr_t *end) {
    if (length == 0) {
        return false;
    }
    if (start > std::numeric_limits<std::uintptr_t>::max() - length) {
        return false;
    }

    *end = start + length;
    return true;
}

static bool is_sv39_canonical(std::uintptr_t vaddr) {
    static constexpr std::uintptr_t SIGN_BIT = static_cast<std::uintptr_t>(1) << 38;
    static constexpr std::uintptr_t LOWER_MASK = (static_cast<std::uintptr_t>(1) << 39) - 1;
    const std::uintptr_t upper = vaddr & ~LOWER_MASK;

    if ((vaddr & SIGN_BIT) == 0) {
        return upper == 0;
    }
    return upper == ~LOWER_MASK;
}

static std::uintptr_t vpn_index(std::uintptr_t vaddr, int level) {
    return (vaddr >> (PAGE_SHIFT + static_cast<std::uintptr_t>(level) * VPN_BITS)) & VPN_MASK;
}

static bool read_pte(const Bus &bus, std::uintptr_t paddr, std::uint64_t *pte) {
    std::uint8_t *host_addr = bus.translate(paddr, PTE_SIZE);
    if (host_addr == nullptr) {
        return false;
    }

    std::uint64_t value = 0;
    std::memcpy(&value, host_addr, sizeof(value));
    *pte = value;
    return true;
}

static bool is_valid_pte(std::uint64_t pte) {
    const bool valid = (pte & PTE_V) != 0;
    const bool readable = (pte & PTE_R) != 0;
    const bool writable = (pte & PTE_W) != 0;

    return valid && (readable || !writable);
}

static bool is_leaf_pte(std::uint64_t pte) {
    return (pte & (PTE_R | PTE_X)) != 0;
}

static std::uintptr_t pte_ppn(std::uint64_t pte) {
    return static_cast<std::uintptr_t>((pte >> 10) & PTE_PPN_MASK);
}

static bool range_fits_in_leaf(std::uintptr_t vaddr, std::uintptr_t length, int level) {
    const std::uintptr_t page_size = PAGE_SIZE << (static_cast<std::uintptr_t>(level) * VPN_BITS);
    const std::uintptr_t page_offset = vaddr & (page_size - 1);
    return length <= page_size - page_offset;
}

std::uint8_t *KvmCpu::translate_guest_vaddr(std::uintptr_t guest_vaddr, std::uintptr_t length) const {
    if (bus_ == nullptr) {
        return nullptr;
    }

    std::uintptr_t end = 0;
    if (!checked_range_end(guest_vaddr, length, &end)) {
        return nullptr;
    }

    KvmSRegs sregs = {};
    if (!get_sregs(sregs)) {
        return nullptr;
    }

    const std::uintptr_t satp_mode = sregs.satp >> SATP_MODE_SHIFT;
    if (satp_mode == SATP_MODE_BARE) {
        return bus_->translate(guest_vaddr, length);
    }
    if (satp_mode != SATP_MODE_SV39) {
        std::fprintf(stderr, "unsupported guest satp mode: 0x%lx\n", static_cast<unsigned long>(satp_mode));
        return nullptr;
    }
    if (!is_sv39_canonical(guest_vaddr) || !is_sv39_canonical(end - 1)) {
        return nullptr;
    }

    std::uintptr_t table_paddr = (sregs.satp & SATP_PPN_MASK) << PAGE_SHIFT;
    for (int level = 2; level >= 0; level--) {
        const std::uintptr_t pte_addr = table_paddr + vpn_index(guest_vaddr, level) * PTE_SIZE;
        std::uint64_t pte = 0;
        if (!read_pte(*bus_, pte_addr, &pte) || !is_valid_pte(pte)) {
            return nullptr;
        }

        const std::uintptr_t ppn = pte_ppn(pte);
        if (is_leaf_pte(pte)) {
            const std::uintptr_t lower_ppn_mask =
                (static_cast<std::uintptr_t>(1) << (static_cast<std::uintptr_t>(level) * VPN_BITS)) - 1;
            if ((ppn & lower_ppn_mask) != 0 || !range_fits_in_leaf(guest_vaddr, length, level)) {
                return nullptr;
            }

            const std::uintptr_t page_shift = PAGE_SHIFT + static_cast<std::uintptr_t>(level) * VPN_BITS;
            const std::uintptr_t page_offset = guest_vaddr & ((static_cast<std::uintptr_t>(1) << page_shift) - 1);
            const std::uintptr_t guest_paddr = (ppn << PAGE_SHIFT) | page_offset;
            return bus_->translate(guest_paddr, length);
        }

        table_paddr = ppn << PAGE_SHIFT;
    }

    return nullptr;
}
} // namespace kvm_host

File: usertests/kvm/host-rs/src/vcpu/pagetable.rs
use std::{mem, ptr};

use num_enum::TryFromPrimitive;

use crate::device::bus::{Bus, checked_range_end};
use crate::vcpu::KvmCpu;

const PAGE_SHIFT: usize = 12;
const PAGE_SIZE: usize = 1 << PAGE_SHIFT;

impl KvmCpu {
    pub fn translate_read(&self, guest_vaddr: usize, buffer: &mut [u8]) -> bool {
        let bus = self.bus();
        let bus = bus.lock().expect("kvm bus lock poisoned");
        let Some(chunks) = self.translate_chunks(&bus, guest_vaddr, buffer.len(), TranslateAccess::Read) else {
            return false;
        };

        for (host_addr, offset, chunk) in chunks {
            unsafe {
                ptr::copy(host_addr, buffer.as_mut_ptr().add(offset), chunk);
            }
        }
        true
    }

    #[allow(dead_code)]
    pub fn translate_write(&self, guest_vaddr: usize, buffer: &[u8]) -> bool {
        let bus = self.bus();
        let bus = bus.lock().expect("kvm bus lock poisoned");
        let Some(chunks) = self.translate_chunks(&bus, guest_vaddr, buffer.len(), TranslateAccess::Write) else {
            return false;
        };

        for (host_addr, offset, chunk) in chunks {
            unsafe {
                ptr::copy(buffer.as_ptr().add(offset), host_addr, chunk);
            }
        }
        true
    }

    fn translate_chunks(
        &self,
        bus: &Bus,
        mut guest_vaddr: usize,
        length: usize,
        access: TranslateAccess,
    ) -> Option<Vec<(*mut u8, usize, usize)>> {
        if length == 0 {
            return Some(Vec::new());
        }
        checked_range_end(guest_vaddr, length)?;
        let satp = self.get_sregs().ok()?.satp;

        let mut offset = 0;
        let mut chunks = Vec::new();
        while offset < length {
            let chunk = guest_page_chunk(guest_vaddr, length - offset);
            let host_addr = self.translate_guest_vaddr(bus, satp, guest_vaddr, chunk, access)?;
            chunks.push((host_addr, offset, chunk));
            guest_vaddr += chunk;
            offset += chunk;
        }
        Some(chunks)
    }

    fn translate_guest_vaddr(
        &self,
        bus: &Bus,
        satp: usize,
        guest_vaddr: usize,
        length: usize,
        access: TranslateAccess,
    ) -> Option<*mut u8> {
        let end = checked_range_end(guest_vaddr, length)?;
        match SatpMode::try_from(satp >> 60).ok()? {
            SatpMode::Bare => return bus.translate(guest_vaddr, length),
            SatpMode::Sv39 => {}
        }
        if !is_sv39_canonical(guest_vaddr) || !is_sv39_canonical(end - 1) {
            return None;
        }

        let mut table_paddr = (satp & ((1usize << 44) - 1)) << 12;
        for level in (0..=2).rev() {
            let pte_addr = table_paddr + vpn_index(guest_vaddr, level) * 8;
            let pte = read_pte(bus, pte_addr)?;
            if !is_valid_pte(pte) {
                return None;
            }
            let ppn = ((pte >> 10) & ((1u64 << 44) - 1)) as usize;
            if is_leaf_pte(pte) {
                if !access.allowed_by(pte) {
                    return None;
                }
                let lower_ppn_mask = (1usize << (level * 9)) - 1;
                if (ppn & lower_ppn_mask) != 0 || !range_fits_in_leaf(guest_vaddr, length, level) {
                    return None;
                }
                let page_shift = 12 + level * 9;
                let page_offset = guest_vaddr & ((1usize << page_shift) - 1);
                let guest_paddr = (ppn << 12) | page_offset;
                return bus.translate(guest_paddr, length);
            }
            table_paddr = ppn << 12;
        }
        None
    }
}

#[derive(Clone, Copy)]
enum TranslateAccess {
    Read,
    Write,
}

impl TranslateAccess {
    fn allowed_by(self, pte: u64) -> bool {
        match self {
            Self::Read => PteFlag::Read.is_set(pte),
            Self::Write => PteFlag::Write.is_set(pte),
        }
    }
}

#[repr(u64)]
#[derive(Clone, Copy)]
enum PteFlag {
    Valid = 1 << 0,
    Read = 1 << 1,
    Write = 1 << 2,
    Execute = 1 << 3,
}

impl PteFlag {
    fn is_set(self, pte: u64) -> bool {
        (pte & self as u64) != 0
    }
}

#[repr(usize)]
#[derive(Clone, Copy, TryFromPrimitive)]
enum SatpMode {
    Bare = 0,
    Sv39 = 8,
}

fn guest_page_chunk(guest_vaddr: usize, remaining: usize) -> usize {
    let page_remaining = PAGE_SIZE - (guest_vaddr & (PAGE_SIZE - 1));
    remaining.min(page_remaining)
}

fn is_sv39_canonical(vaddr: usize) -> bool {
    let sign_bit = 1usize << 38;
    let lower_mask = (1usize << 39) - 1;
    let upper = vaddr & !lower_mask;
    if (vaddr & sign_bit) == 0 {
        upper == 0
    } else {
        upper == !lower_mask
    }
}

fn vpn_index(vaddr: usize, level: usize) -> usize {
    (vaddr >> (PAGE_SHIFT + level * 9)) & ((1 << 9) - 1)
}

fn read_pte(bus: &Bus, paddr: usize) -> Option<u64> {
    let host_addr = bus.translate(paddr, mem::size_of::<u64>())?;
    let mut value = 0u64;
    unsafe {
        ptr::copy_nonoverlapping(host_addr, &mut value as *mut u64 as *mut u8, mem::size_of::<u64>());
    }
    Some(value)
}

fn is_valid_pte(pte: u64) -> bool {
    let valid = PteFlag::Valid.is_set(pte);
    let readable = PteFlag::Read.is_set(pte);
    let writable = PteFlag::Write.is_set(pte);
    valid && (readable || !writable)
}

fn is_leaf_pte(pte: u64) -> bool {
    PteFlag::Read.is_set(pte) || PteFlag::Execute.is_set(pte)
}

fn range_fits_in_leaf(vaddr: usize, length: usize, level: usize) -> bool {
    let page_size = 1usize << (PAGE_SHIFT + level * 9);
    let page_offset = vaddr & (page_size - 1);
    length <= page_size - page_offset
}
